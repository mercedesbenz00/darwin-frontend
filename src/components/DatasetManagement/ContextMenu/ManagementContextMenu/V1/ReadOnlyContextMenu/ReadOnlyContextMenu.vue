<template>
  <gallery-context-menu v-if="isSelecting">
    <gallery-context-menu-item
      class="readonly-context-menu__selected"
      :label="`${selectedCount} Selected`"
    />
  </gallery-context-menu>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import { State } from 'vuex-class'

import GalleryContextMenu from '@/components/Common/Gallery/GalleryContextMenu.vue'
import GalleryContextMenuItem from '@/components/Common/Gallery/GalleryContextMenuItem.vue'
import { DatasetItemPayload } from '@/store/types'

/**
 * Used to render a context menu in open datasets (base and video).
 */
@Component({
  name: 'workflow-read-only-context-menu',
  components: { GalleryContextMenu, GalleryContextMenuItem }
})
export default class ReadOnlyContextMenu extends Vue {
  @State(state => state.dataset.datasetItems)
  datasetItems!: DatasetItemPayload[]

  @State(state => state.dataset.selectedItemIds)
  selectedItemIds!: number[]

  /**
   * Returns the currently selected items.
   *
   * Can be used by parent components through a reference.
   */
  get selectedItems (): DatasetItemPayload[] {
    return this.datasetItems.filter(i => this.selectedItemIds.includes(i.id))
  }

  get selectedCount (): number {
    return this.selectedItems.length
  }

  get isSelecting (): boolean {
    return this.selectedCount > 0
  }
}
</script>

<style lang="scss" scoped>
.readonly-context-menu__selected {
  width: 190px;
}
</style>
