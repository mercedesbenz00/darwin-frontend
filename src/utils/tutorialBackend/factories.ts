import {
  AnnotateStagePayload,
  CodeStagePayload,
  CompleteStagePayload,
  DatasetImagePayload,
  DatasetItemPayload,
  DatasetItemStatus,
  DatasetItemType,
  ImagePayload,
  ModelStagePayload,
  ReviewStagePayload,
  StageType,
  WorkflowStagePayload,
  WorkflowStageTemplatePayload,
  WorkflowTemplatePayload
} from '@/store/types'

import TutorialError from './TutorialError'
import { dataset } from './data/dataset'

const itemId = (datasetImageId: number) => datasetImageId * 10
const workflowId = (datasetImageId: number) => datasetImageId * 100

const stageId = (datasetImageId: number, stageNumber: number) =>
  workflowId(datasetImageId) + stageNumber

const initializeDatasetImage = (image: ImagePayload): DatasetImagePayload => ({
  dataset_id: dataset.id,
  dataset_video_id: null,
  id: image.id * 10,
  image,
  seq: image.id,
  set: 1
})

const initializeItem = (datasetImage: DatasetImagePayload): DatasetItemPayload => ({
  archived_reason: null,
  archived: false,
  current_workflow: null,
  current_workflow_id: null,
  dataset_id: datasetImage.dataset_id,
  dataset_image_id: datasetImage.id,
  dataset_image: datasetImage,
  dataset_video_id: null,
  dataset_video: null,
  file_size: 0,
  filename: datasetImage.image.original_filename,
  height: datasetImage.image.height,
  id: itemId(datasetImage.id),
  inserted_at: '2020-06-12T00:00:00',
  labels: [],
  path: '/',
  priority: 0,
  seq: datasetImage.seq,
  set: datasetImage.set,
  status: DatasetItemStatus.annotate,
  updated_at: '2020-06-12T00:00:00',
  width: datasetImage.image.width,
  type: DatasetItemType.image
})

const initializeStage = (
  datasetImage: DatasetImagePayload,
  stageTemplate: WorkflowStageTemplatePayload
): WorkflowStagePayload => {
  const { type } = stageTemplate
  const payload = {
    assignee_id: null,
    completed: false,
    completes_at: null,
    dataset_item_id: itemId(datasetImage.id),
    id: stageId(datasetImage.id, stageTemplate.stage_number),
    metadata: {},
    number: stageTemplate.stage_number,
    skipped_reason: null,
    skipped: false,
    template_metadata: stageTemplate.metadata,
    type: stageTemplate.type,
    workflow_id: workflowId(datasetImage.id),
    workflow_stage_template_id: stageTemplate.id
  }

  if (type === StageType.Annotate) { return payload as AnnotateStagePayload }
  if (type === StageType.Code) { return payload as CodeStagePayload }
  if (type === StageType.Complete) { return payload as CompleteStagePayload }
  if (type === StageType.Model) { return payload as ModelStagePayload }
  if (type === StageType.Review) { return payload as ReviewStagePayload }

  throw new Error('Invalid type given')
}

const initializeWorkflow = (
  datasetImage: DatasetImagePayload,
  workflowTemplate: WorkflowTemplatePayload,
  currentStage: number
): Pick<DatasetItemPayload, 'current_workflow' | 'current_workflow_id'> => {
  const annotateTemplate =
    workflowTemplate.workflow_stage_templates.find(t => t.type === StageType.Annotate)

  const reviewTemplate =
    workflowTemplate.workflow_stage_templates.find(t => t.type === StageType.Review)

  const completeTemplate =
    workflowTemplate.workflow_stage_templates.find(t => t.type === StageType.Complete)

  if (!annotateTemplate || !reviewTemplate || !completeTemplate) {
    throw new TutorialError('Initializing workflow using invalid template ')
  }

  const id = workflowId(datasetImage.id)

  const workflow: DatasetItemPayload['current_workflow'] = {
    dataset_item_id: itemId(datasetImage.id),
    id,
    workflow_template_id: workflowTemplate.id,
    current_stage_number: currentStage,
    current_workflow_stage_template_id: annotateTemplate.id,
    status: DatasetItemStatus.annotate,
    stages: {
      1: [initializeStage(datasetImage, annotateTemplate)],
      2: [initializeStage(datasetImage, reviewTemplate)],
      3: [initializeStage(datasetImage, completeTemplate)]
    }
  }

  return { current_workflow: workflow, current_workflow_id: id }
}

export const createItem = (
  image: ImagePayload,
  workflowTeplate: WorkflowTemplatePayload,
  stageNumber: number = 1
): DatasetItemPayload => {
  const datasetImage = initializeDatasetImage(image)
  return {
    ...initializeItem(datasetImage),
    ...initializeWorkflow(datasetImage, workflowTeplate, stageNumber)
  }
}
