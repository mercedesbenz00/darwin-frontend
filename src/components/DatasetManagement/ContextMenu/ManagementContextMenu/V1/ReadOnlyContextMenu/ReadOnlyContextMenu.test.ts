import { createLocalVue, shallowMount } from '@vue/test-utils'
import Vuex from 'vuex'

import createTestStore from 'test/unit/createTestStore'
import { buildDatasetItemPayload } from 'test/unit/factories'

import ReadOnlyContextMenu from './ReadOnlyContextMenu.vue'

const localVue = createLocalVue()
localVue.use(Vuex)

let store: ReturnType<typeof createTestStore>

beforeEach(() => {
  store = createTestStore()
})

const items = [
  buildDatasetItemPayload({ id: 1, seq: 1, dataset_id: 1 }),
  buildDatasetItemPayload({ id: 2, seq: 2, dataset_id: 1 }),
  buildDatasetItemPayload({ id: 3, seq: 3, dataset_id: 1 })
]

it('matches snapshot when nothing selected', () => {
  store.commit('dataset/SET_DATASET_ITEMS', items)
  const wrapper = shallowMount(ReadOnlyContextMenu, { localVue, store })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot with selection', () => {
  store.commit('dataset/SET_DATASET_ITEMS', items)
  store.commit('dataset/UPDATE_ITEM_SELECTION', { items: items.slice(0, 2), selected: true })
  const wrapper = shallowMount(ReadOnlyContextMenu, { localVue, store })
  expect(wrapper).toMatchSnapshot()
})
