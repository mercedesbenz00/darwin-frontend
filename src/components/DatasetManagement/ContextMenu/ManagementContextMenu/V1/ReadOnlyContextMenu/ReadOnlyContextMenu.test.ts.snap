// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot when nothing selected 1`] = ``;

exports[`matches snapshot with selection 1`] = `
<gallery-context-menu-stub>
  <gallery-context-menu-item-stub label="2 Selected" class="readonly-context-menu__selected"></gallery-context-menu-item-stub>
</gallery-context-menu-stub>
`;
